### PR TITLE
[DirectX] Add DXIL_MODULE_FLAG for ShaderFlags.

### LINKED ISSUE #57925: [DirectX backend] Build shader flag which saved into DXIL entry metadata
The layout is 64bit unsigned integer union with

    unsigned m_bDisableOptimizations :1;   // D3D11_1_SB_GLOBAL_FLAG_SKIP_OPTIMIZATION
    unsigned m_bDisableMathRefactoring :1; //~D3D10_SB_GLOBAL_FLAG_REFACTORING_ALLOWED
    unsigned m_bEnableDoublePrecision :1; // D3D11_SB_GLOBAL_FLAG_ENABLE_DOUBLE_PRECISION_FLOAT_OPS
    unsigned m_bForceEarlyDepthStencil :1; // D3D11_SB_GLOBAL_FLAG_FORCE_EARLY_DEPTH_STENCIL
    unsigned m_bEnableRawAndStructuredBuffers :1; // D3D11_SB_GLOBAL_FLAG_ENABLE_RAW_AND_STRUCTURED_BUFFERS
    unsigned m_bLowPrecisionPresent :1; // D3D11_1_SB_GLOBAL_FLAG_ENABLE_MINIMUM_PRECISION
    unsigned m_bEnableDoubleExtensions :1; // D3D11_1_SB_GLOBAL_FLAG_ENABLE_DOUBLE_EXTENSIONS
    unsigned m_bEnableMSAD :1;        // D3D11_1_SB_GLOBAL_FLAG_ENABLE_SHADER_EXTENSIONS
    unsigned m_bAllResourcesBound :1; // D3D12_SB_GLOBAL_FLAG_ALL_RESOURCES_BOUND

    unsigned m_bViewportAndRTArrayIndex :1;   // SHADER_FEATURE_VIEWPORT_AND_RT_ARRAY_INDEX_FROM_ANY_SHADER_FEEDING_RASTERIZER
    unsigned m_bInnerCoverage :1;             // SHADER_FEATURE_INNER_COVERAGE
    unsigned m_bStencilRef  :1;               // SHADER_FEATURE_STENCIL_REF
    unsigned m_bTiledResources  :1;           // SHADER_FEATURE_TILED_RESOURCES
    unsigned m_bUAVLoadAdditionalFormats :1;  // SHADER_FEATURE_TYPED_UAV_LOAD_ADDITIONAL_FORMATS
    unsigned m_bLevel9ComparisonFiltering :1; // SHADER_FEATURE_LEVEL_9_COMPARISON_FILTERING
                                              // SHADER_FEATURE_11_1_SHADER_EXTENSIONS shared with EnableMSAD
    unsigned m_b64UAVs :1;                    // SHADER_FEATURE_64_UAVS
    unsigned m_UAVsAtEveryStage :1;           // SHADER_FEATURE_UAVS_AT_EVERY_STAGE
    unsigned m_bCSRawAndStructuredViaShader4X : 1; // SHADER_FEATURE_COMPUTE_SHADERS_PLUS_RAW_AND_STRUCTURED_BUFFERS_VIA_SHADER_4_X
    
    // SHADER_FEATURE_COMPUTE_SHADERS_PLUS_RAW_AND_STRUCTURED_BUFFERS_VIA_SHADER_4_X is specifically
    // about shader model 4.x.

    unsigned m_bROVS :1;              // SHADER_FEATURE_ROVS
    unsigned m_bWaveOps :1;           // SHADER_FEATURE_WAVE_OPS
    unsigned m_bInt64Ops :1;          // SHADER_FEATURE_INT64_OPS
    unsigned m_bViewID : 1;           // SHADER_FEATURE_VIEWID
    unsigned m_bBarycentrics : 1;     // SHADER_FEATURE_BARYCENTRICS

    unsigned m_bUseNativeLowPrecision : 1;

    unsigned m_bShadingRate : 1;      // SHADER_FEATURE_SHADINGRATE

    unsigned m_bRaytracingTier1_1 : 1; // SHADER_FEATURE_RAYTRACING_TIER_1_1
    unsigned m_bSamplerFeedback : 1; // SHADER_FEATURE_SAMPLER_FEEDBACK

    unsigned m_bAtomicInt64OnTypedResource : 1; // SHADER_FEATURE_ATOMIC_INT64_ON_TYPED_RESOURCE
    unsigned m_bAtomicInt64OnGroupShared : 1; // SHADER_FEATURE_ATOMIC_INT64_ON_GROUP_SHARED

    unsigned m_bDerivativesInMeshAndAmpShaders : 1; //SHADER_FEATURE_DERIVATIVES_IN_MESH_AND_AMPLIFICATION_SHADERS

    unsigned m_bResourceDescriptorHeapIndexing : 1;  // SHADER_FEATURE_RESOURCE_DESCRIPTOR_HEAP_INDEXING
    unsigned m_bSamplerDescriptorHeapIndexing : 1;  // SHADER_FEATURE_SAMPLER_DESCRIPTOR_HEAP_INDEXING

    unsigned m_bAtomicInt64OnHeapResource : 1; // SHADER_FEATURE_ATOMIC_INT64_ON_DESCRIPTOR_HEAP_RESOURCE

    // Global flag indicating that any UAV may not alias any other UAV.
    // Set if UAVs are used, unless -res-may-alias was specified.
    // For modules compiled against validator version < 1.7, this flag will be
    // cleared, and it must be assumed that UAV resources may alias.
    unsigned m_bResMayNotAlias : 1;

    unsigned m_bAdvancedTextureOps : 1;     // SHADER_FEATURE_ADVANCED_TEXTURE_OPS
    unsigned m_bWriteableMSAATextures : 1;  // SHADER_FEATURE_WRITEABLE_MSAA_TEXTURES

    uint32_t m_align1 : 28;            // align to 64 bit.

### PR DESCRIPTION
**Pull request by python3kgae**
Add DXILConstants.def which stores DXILModuleFlags.

Use DXILModuleFlags for ComputedShaderFlags instead of ShaderFeatureFlags.

ComputedShaderFlags::getFeatureInfo() was added to get FeatureInfoFlags.

Fixes #57925

### CODE REVIEW

File: llvm/include/llvm/BinaryFormat/DXContainerConstants.def [hunk 1]
```diff
@@ -44,8 +44,11 @@ SHADER_FEATURE_FLAG(27, RESERVED, "<RESERVED>")
 SHADER_FEATURE_FLAG(28, AtomicInt64OnHeapResource, "64-bit Atomics on Heap Resources")
 SHADER_FEATURE_FLAG(29, AdvancedTextureOps, "Advanced Texture Ops")
 SHADER_FEATURE_FLAG(30, WriteableMSAATextures, "Writeable MSAA Textures")
+SHADER_FEATURE_FLAG(31, WaveMMA, "Wave Matrix Multiply Accumulate")
+SHADER_FEATURE_FLAG(32, SampleCmpGradientOrBias, "SampleCmpGradient or SampleCmpLevelZeroBias")
+SHADER_FEATURE_FLAG(33, ExtendedCommandInfo, "StartVertexLocation and StartInstanceLocation")
```
Comment (llvm-beanz) [inline llvm/include/llvm/BinaryFormat/DXContainerConstants.def:49]: Adding these flag values is tracked in a separate task (https://github.com/llvm/llvm-project/issues/83177) and shouldn't be included in this PR.

File: llvm/include/llvm/Support/DXILConstants.def [hunk 2]
```diff
@@ -0,0 +1,76 @@
+
+#ifdef DXIL_MODULE_FLAG
+
+// DXIL_MODULE_FLAG(bit offset for the flag, bit offset for feature flag (-1 for no feature flag), name, description.
+DXIL_MODULE_FLAG( 0, -1,  DisableOptimizations,   "D3D11_1_SB_GLOBAL_FLAG_SKIP_OPTIMIZATION")
+DXIL_MODULE_FLAG( 1, -1,  DisableMathRefactoring, "D3D10_SB_GLOBAL_FLAG_REFACTORING_ALLOWED")
+DXIL_MODULE_FLAG( 2,  0,  EnableDoublePrecision,  "Double-precision floating point")
```
Comment (llvm-beanz) [inline llvm/include/llvm/Support/DXILConstants.def:7]: A bunch of these flags are duplicated in the DXContainerCosntants. Is there a reason we need to define them twice?

File: llvm/lib/Target/DirectX/DXContainerGlobals.cpp [hunk 3]
```diff
@@ -28,7 +28,7 @@ using namespace llvm::dxil;
 namespace {
 class DXContainerGlobals : public llvm::ModulePass {
 
-  GlobalVariable *getShaderFlags(Module &M);
+  GlobalVariable *getShaderFeatureInfo(Module &M);
```
Comment (llvm-beanz) [inline llvm/lib/Target/DirectX/DXContainerGlobals.cpp:31]: This name is confusing. Why are we calling a flag encoding "feature info"?

File: llvm/test/CodeGen/DirectX/ShaderFlags/double-extensions.ll [hunk 4]
```diff
@@ -3,7 +3,7 @@
 
 target triple = "dxil-pc-shadermodel6.7-library"
 
-; CHECK: ; Shader Flags Value: 0x00000021
+; CHECK: ; Shader Flags Value: 0x00000044
```
Comment (llvm-beanz) [inline llvm/test/CodeGen/DirectX/ShaderFlags/double-extensions.ll:6]: Do the extra flags get pretty-printed below? If not, should they? Should we check them?

File: llvm/include/llvm/Support/DXILConstants.def (hunk 2 above)
Comment (python3kgae) [reply llvm/include/llvm/Support/DXILConstants.def:7]: We could share, just need to add a field to SHADER_FEATURE_FLAG for DXILModuleFlag bit offset.
If we put all things as SHADER_FEATURE_FLAG, there will be something not SHADER_FEATURE_FLAG though.

File: llvm/test/CodeGen/DirectX/ShaderFlags/double-extensions.ll (hunk 4 above)
Comment (python3kgae) [reply llvm/test/CodeGen/DirectX/ShaderFlags/double-extensions.ll:6]: The following CHECK lines are the pretty-printed.

File: llvm/lib/Target/DirectX/DXContainerGlobals.cpp (hunk 3 above)
Comment (python3kgae) [reply llvm/lib/Target/DirectX/DXContainerGlobals.cpp:31]: So just getFeatureInfo(Module &M)?

File: llvm/include/llvm/BinaryFormat/DXContainerConstants.def (hunk 1 above)
Comment (python3kgae) [reply llvm/include/llvm/BinaryFormat/DXContainerConstants.def:49]: I'll remove these changes.

File: llvm/test/CodeGen/DirectX/ShaderFlags/double-extensions.ll (hunk 4 above)
Comment (llvm-beanz) [reply llvm/test/CodeGen/DirectX/ShaderFlags/double-extensions.ll:6]: There are new flags you've added which don't have tests verifying the pretty printing.

Comment (python3kgae) [reply llvm/test/CodeGen/DirectX/ShaderFlags/double-extensions.ll:6]: We don't have all the required dxil operation or compiler option added for the full list of DxilModuleFlags. If a DxilModuleFlag is not set, it will not be printed.
Test will be added when we really set these flags.

Comment (python3kgae) [reply llvm/test/CodeGen/DirectX/ShaderFlags/double-extensions.ll:6]: For double flag, which already supported.
The print check is in the following CHECK lines as
; CHECK-NEXT: ;       Double-precision floating point
; CHECK-NEXT: ;       Double-precision extensions for 11.1

We don't support other flags right now.
We don't have dxil intrinsic like waveOps.

File: llvm/lib/Target/DirectX/DXILShaderFlags.h [hunk 5]
```diff
@@ -29,23 +28,35 @@ class GlobalVariable;
 namespace dxil {
 
 struct ComputedShaderFlags {
-#define SHADER_FEATURE_FLAG(bit, FlagName, Str) bool FlagName : 1;
+#define SHADER_FEATURE_FLAG(bit, DxilModuleBit, FlagName, Str)                 \
```
Comment (pow2clk) [inline llvm/lib/Target/DirectX/DXILShaderFlags.h:31]: The name of this parameter changes here, which is a little confusing

File: llvm/lib/Target/DirectX/DXILShaderFlags.cpp [hunk 6]
```diff
@@ -51,7 +51,10 @@ void ComputedShaderFlags::print(raw_ostream &OS) const {
   if (FlagVal == 0)
     return;
   OS << "; Note: shader requires additional functionality:\n";
-#define SHADER_FEATURE_FLAG(bit, FlagName, Str)                                \
+#define SHADER_FEATURE_FLAG(bit, DxilModuleNum, FlagName, Str)                 \
```
Comment (python3kgae) [inline llvm/lib/Target/DirectX/DXILShaderFlags.cpp:54]:
```suggestion
#define SHADER_FEATURE_FLAG(bit, DxilModuleBit, FlagName, Str)                 \
```

File: llvm/test/CodeGen/DirectX/ShaderFlags/double-extensions.ll (hunk 4 above)
Comment (llvm-beanz) [reply llvm/test/CodeGen/DirectX/ShaderFlags/double-extensions.ll:6]: Repeating the same thing back to me doesn't help me understand the issue here.

I went and looked at the code, and I think we have a separate issue we need to resolve, but that can be separate. We should be able to dump the flags from the computed value that is stored in the module, but today the printer only operates on the analysis result. We can address that separately.

File: llvm/lib/Target/DirectX/DXContainerGlobals.cpp (hunk 3 above)
Comment (llvm-beanz) [reply llvm/lib/Target/DirectX/DXContainerGlobals.cpp]: "feature info" doesn't mean anything to me when I read the code. The GlobalVaraible returned here, and the `uint64_t` returned in other places they aren't a "feature info". They are "feature flags".

File: llvm/lib/Target/DirectX/DXILShaderFlags.cpp (hunk 6 above)
Comment (llvm-beanz) [inline llvm/lib/Target/DirectX/DXILShaderFlags.cpp:54]: Your suggestion is good, but also please fix the capitalization on the first argument:

```suggestion
#define SHADER_FEATURE_FLAG(Bit, DxilModuleBit, FlagName, Str)                 \
```

File: llvm/lib/Target/DirectX/DXILShaderFlags.cpp [hunk 7]
```diff
@@ -51,7 +51,10 @@ void ComputedShaderFlags::print(raw_ostream &OS) const {
...
+  if (FlagName)                                                                \
+    OS << ";       " Str "\n";
+#define DXIL_MODULE_FLAG(bit, FlagName, Str)                                   \
```
Comment (llvm-beanz) [inline llvm/lib/Target/DirectX/DXILShaderFlags.cpp:57]:
```suggestion
#define DXIL_MODULE_FLAG(Bit, FlagName, Str)                                   \
```

File: llvm/lib/Target/DirectX/DXContainerGlobals.cpp (hunk 3 above)
Comment (python3kgae) [reply llvm/lib/Target/DirectX/DXContainerGlobals.cpp]: Changed to getFeatureFlags.

File: llvm/lib/Target/DirectX/DXILShaderFlags.cpp (hunk 6 above)
Comment (python3kgae) [reply llvm/lib/Target/DirectX/DXILShaderFlags.cpp]: Fixed.

File: llvm/lib/Target/DirectX/DXILShaderFlags.cpp (hunk 7 above)
Comment (python3kgae) [reply llvm/lib/Target/DirectX/DXILShaderFlags.cpp]: Fixed.

File: llvm/lib/Target/DirectX/DXILShaderFlags.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (python3kgae) [reply llvm/lib/Target/DirectX/DXILShaderFlags.h]: Fixed.

**Review by llvm-beanz — APPROVED** [review]

File: llvm/include/llvm/BinaryFormat/DXContainerConstants.def [hunk 8]
```diff
@@ -13,43 +13,58 @@ CONTAINER_PART(PSG1)
...
-SHADER_FEATURE_FLAG(24, DerivativesInMeshAndAmpShaders, "Derivatives in mesh and amplification shaders")
-SHADER_FEATURE_FLAG(25, ResourceDescriptorHeapIndexing, "Resource descriptor heap indexing")
-SHADER_FEATURE_FLAG(26, SamplerDescriptorHeapIndexing, "Sampler descriptor heap indexing")
-SHADER_FEATURE_FLAG(27, RESERVED, "<RESERVED>")
-SHADER_FEATURE_FLAG(28, AtomicInt64OnHeapResource, "64-bit Atomics on Heap Resources")
-SHADER_FEATURE_FLAG(29, AdvancedTextureOps, "Advanced Texture Ops")
-SHADER_FEATURE_FLAG(30, WriteableMSAATextures, "Writeable MSAA Textures")
-
-SHADER_FEATURE_FLAG(31, NextUnusedBit, "Next reserved shader flag bit (not a flag)")
+// SHADER_FEATURE_FLAG(bit offset for the flag, bit offset for DXIL module flag, name, description.
```
Comment (tex3d) [inline llvm/include/llvm/BinaryFormat/DXContainerConstants.def:16]: `bit offset for the flag` - in what context?  This is the offset in the Shader Feature Info flags (in `SFI0` part) or per-function `FeatureInfo*` flags (in `RDAT` part).  Should this read `bit offset for the shader feature info flag`?

Next, I'm not sure `DXIL module flag` is the right term for all of these, since most of them are computed on a per-function or shader basis, and the resulting feature flags are preserved per-function for libraries in `RDAT`.  These are unioned into module-level shader flags, but they aren't all strictly module flags.  These are previously referred to as the DXIL Shader Flags, with the 64-bit flags value referred to as the raw shader flags.

Some of the flags *are* module-scope flags though, so it would seem we need to clarify the distinctions, or this will just be more confusing.  It would seem important to preserve true module-level flags/settings from CodeGen on, separately from the flags computed based on function usage later.  We should have a term for this subset of flags, separate from the raw shader flags, and something like DXIL module flags or DXIL module settings (if there are more than just flags) would seem appropriate for these.  Grouping them can make it easier to check for consistent settings between modules when linking, for instance.

These flags are part of the shader interface:
- Feature Flags, aka. Shader Feature Info flags encoded into the container's `SFI0` part, or per-function into the `RDAT` part.  This is part of the runtime interface, used to validate required feature support at runtime.  Many, but not all, of these flags have a 1:1 mapping from DxilShaderFlags.
- DXIL Shader Flags: all the flags including module flags and per-function/shader computed flags indicating feature usage.  This is a 64-bit value that is part of the driver interface.  This is what's being referred to as DXIL module flags here.

You could think of DxilShaderFlags as breaking down into:
- Global module flags indicating module interpretation, or allowed optimizations, such as `UseNativeLowPrecision`, `AllResourcesBound`, `DisableMathRefactoring`, etc...  These are set based on compilation options.
- Shader flags computed from functions/entry points or global state (or ComputedShaderFlags):
  - most indicating a required feature flag like most of the flags defined with `SHADER_FEATURE_FLAG` here
  - some potentially indicating something for the driver, without mapping to a required feature bit, such as `ForceEarlyDepthStencil`
  - others potentially indicating usage of a feature in a function that may restrict the context from which that function may be called.

I don't think this explanation gives a definitive answer to how we should clarify our flags terminology and organization, but I don't think we have it quite right at this time.

File: llvm/include/llvm/BinaryFormat/DXContainerConstants.def [hunk 9]
```diff
@@ -13,43 +13,58 @@ CONTAINER_PART(PSG1)
...
 
 #undef SHADER_FEATURE_FLAG
 #endif
 
+#ifdef DXIL_MODULE_FLAG
+
+// Only save DXIL module flags which not map to feature flags here.
+DXIL_MODULE_FLAG( 0,  DisableOptimizations,   "D3D11_1_SB_GLOBAL_FLAG_SKIP_OPTIMIZATION")
+DXIL_MODULE_FLAG( 1,  DisableMathRefactoring, "D3D10_SB_GLOBAL_FLAG_REFACTORING_ALLOWED")
+DXIL_MODULE_FLAG( 3,  ForceEarlyDepthStencil, "D3D11_SB_GLOBAL_FLAG_FORCE_EARLY_DEPTH_STENCIL")
```
Comment (tex3d) [inline llvm/include/llvm/BinaryFormat/DXContainerConstants.def:60]: The `*_SB_GLOBAL_FLAG_*` values were carried over from DXBC, and were global for a particular shader.  Most of these can be considered global module-level flags, but not `ForceEarlyDepthStencil`, which applies to a particular pixel shader only (which only makes a difference for libraries).  However, we could also potentially support some at a more granular level, such as `DisableOptimizations` or `DisableMathRefactoring`, though more granular settings for these could be expressed differently.

That said, there are ones missing from this list, and there may be overlap between flags that could be exposed through feature flags or global flags corresponding to DXBC's `*_SB_GLOBAL_FLAG_*`, such as `DX11_1_ShaderExtensions`.

File: llvm/lib/Target/DirectX/DXILShaderFlags.cpp [hunk 10]
```diff
@@ -51,7 +51,10 @@ void ComputedShaderFlags::print(raw_ostream &OS) const {
   if (FlagVal == 0)
     return;
   OS << "; Note: shader requires additional functionality:\n";
-#define SHADER_FEATURE_FLAG(bit, FlagName, Str)                                \
+#define SHADER_FEATURE_FLAG(Bit, DxilModuleNum, FlagName, Str)                 \
+  if (FlagName)                                                                \
+    OS << ";       " Str "\n";
+#define DXIL_MODULE_FLAG(Bit, FlagName, Str)                                   \
```
Comment (tex3d) [inline llvm/lib/Target/DirectX/DXILShaderFlags.cpp:57]: I don't think module flags should be listed as additional functionality required, but it could still be useful to list them, without implying that they are optional feature requirements.

File: llvm/include/llvm/BinaryFormat/DXContainerConstants.def [hunk 11]
```diff
@@ -13,43 +13,58 @@ CONTAINER_PART(PSG1)
...
+DXIL_MODULE_FLAG( 8,  AllResourcesBound, "D3D12_SB_GLOBAL_FLAG_ALL_RESOURCES_BOUND")
+DXIL_MODULE_FLAG(17,  CSRawAndStructuredViaShader4X, "SHADER_FEATURE_COMPUTE_SHADERS_PLUS_RAW_AND_STRUCTURED_BUFFERS_VIA_SHADER_4_X")
```
Comment (tex3d) [inline llvm/include/llvm/BinaryFormat/DXContainerConstants.def:62]: `CSRawAndStructuredViaShader4X` is a shader feature flag, isn't it?  What's it doing here?  Did this get mixed up with the `EnableRawAndStructuredBuffers` flag corresponding to `D3D11_SB_GLOBAL_FLAG_ENABLE_RAW_AND_STRUCTURED_BUFFERS` that belongs here instead?

File: llvm/include/llvm/BinaryFormat/DXContainerConstants.def [hunk 12]
```diff
@@ -13,43 +13,58 @@ CONTAINER_PART(PSG1)
...
-SHADER_FEATURE_FLAG(27, RESERVED, "<RESERVED>")
-SHADER_FEATURE_FLAG(28, AtomicInt64OnHeapResource, "64-bit Atomics on Heap Resources")
-SHADER_FEATURE_FLAG(29, AdvancedTextureOps, "Advanced Texture Ops")
-SHADER_FEATURE_FLAG(30, WriteableMSAATextures, "Writeable MSAA Textures")
-
-SHADER_FEATURE_FLAG(31, NextUnusedBit, "Next reserved shader flag bit (not a flag)")
+// SHADER_FEATURE_FLAG(bit offset for the flag, bit offset for DXIL module flag, name, description.
+
+SHADER_FEATURE_FLAG(0,   2, Doubles, "Double-precision floating point")
+SHADER_FEATURE_FLAG(1,   4, ComputeShadersPlusRawAndStructuredBuffers, "Raw and Structured buffers")
```
Comment (tex3d) [inline llvm/include/llvm/BinaryFormat/DXContainerConstants.def:19]: This should be the feature flag `CSRawAndStructuredViaShader4X` that's mistakenly listed under global flags instead.

**Review by tex3d — COMMENTED** [review]
I had some thoughts on how these flags are broken down currently, plus some issues with specific flags.

File: llvm/include/llvm/BinaryFormat/DXContainerConstants.def (hunk 8 above)
Comment (python3kgae) [reply llvm/include/llvm/BinaryFormat/DXContainerConstants.def:16]: How about adding something like FlagKind to these SHADER_FEATURE_FLAG/DXIL_SHADER_FLAG ?
The kind could be FEATURE_INFO, GLOBAL_OPTION or FUNCTION.

File: llvm/include/llvm/BinaryFormat/DXContainerConstants.def [hunk 13]
```diff
@@ -13,43 +13,58 @@ CONTAINER_PART(PSG1)
...
+SHADER_FEATURE_FLAG(9,  11, StencilRef, "PS Output Stencil Ref")
+SHADER_FEATURE_FLAG(10, 10, InnerCoverage, "PS Inner Coverage")
+SHADER_FEATURE_FLAG(11, 13, TypedUAVLoadAdditionalFormats, "Typed UAV Load Additional Formats")
+SHADER_FEATURE_FLAG(12, 18, ROVs, "Raster Ordered UAVs")
+SHADER_FEATURE_FLAG(13,  9, ViewportAndRTArrayIndexFromAnyShaderFeedingRasterizer, "SV_RenderTargetArrayIndex or SV_ViewportArrayIndex from any shader feeding rasterizer")
+SHADER_FEATURE_FLAG(14, 19, WaveOps, "Wave level operations")
+SHADER_FEATURE_FLAG(15, 20, Int64Ops, "64-Bit integer")
+SHADER_FEATURE_FLAG(16, 21, ViewID, "View Instancing")
+SHADER_FEATURE_FLAG(17, 22, Barycentrics, "Barycentrics")
+SHADER_FEATURE_FLAG(18, 23, NativeLowPrecision, "Use native low precision")
```
[the remark this answers is not in the document]
Comment (python3kgae) [reply llvm/include/llvm/BinaryFormat/DXContainerConstants.def]: Added separate DXIL_MODULE_FLAG.

File: llvm/lib/Target/DirectX/DXILShaderFlags.h [hunk 14]
```diff
@@ -29,23 +28,35 @@ class GlobalVariable;
 namespace dxil {
 
 struct ComputedShaderFlags {
-#define SHADER_FEATURE_FLAG(bit, FlagName, Str) bool FlagName : 1;
+#define SHADER_FEATURE_FLAG(Bit, DxilModuleBit, FlagName, Str)                 \
+  bool FlagName : 1;
+#define DXIL_MODULE_FLAG(Bit, FlagName, Str) bool FlagName : 1;
```
[the remark this answers is not in the document]
Comment (python3kgae) [reply llvm/lib/Target/DirectX/DXILShaderFlags.h]: Updated

File: llvm/lib/Target/DirectX/DXILShaderFlags.cpp (hunk 10 above)
Comment (python3kgae) [reply llvm/lib/Target/DirectX/DXILShaderFlags.cpp]: Added

File: llvm/include/llvm/BinaryFormat/DXContainerConstants.def (hunk 11 above)
Comment (python3kgae) [reply llvm/include/llvm/BinaryFormat/DXContainerConstants.def]: Good catch.
Fixed.

File: llvm/include/llvm/BinaryFormat/DXContainerConstants.def (hunk 12 above)
Comment (python3kgae) [reply llvm/include/llvm/BinaryFormat/DXContainerConstants.def]: Fixed.

File: llvm/include/llvm/BinaryFormat/DXContainerConstants.def (hunk 8 above)
Comment (python3kgae) [reply llvm/include/llvm/BinaryFormat/DXContainerConstants.def]: Will adding a FlagKind for SHADER_FEATURE_FLAG/DXIL_MODULE_FLAG make things clearer and solve the problem?

Comment (tex3d) [reply llvm/include/llvm/BinaryFormat/DXContainerConstants.def]: I'm not sure that's solving the issue.  I'm trying to think about a breakdown design that results in consistent terminology, and exposes the important distinctions.

What do you think of this?

Terminology:
1. "DXIL Module Flags" - flag values used in DXIL metadata - could these be called "DXIL Shader Flags"?
2. "Shader Feature Flags" - flags used in shader feature info (SFI0) part and FeatureInfo fields in RDAT part - Could we call these "Shader Feature Info Flags"?  It's a bit longer, but clarifies their connection with something more specific.
3. Option flags - we could refer to these simply as "Option Flags".  These are a subset of "DXIL Shader Flags".
4. Computed flags - we could refer to these as "Computed Shader Flags".  These are a subset of "DXIL Shader Flags", most of which are reflected directly into Shader Feature Info flags, some indirectly (like min-precision), and some not at all.

With this terminology, we could use three different macros for defining the flags:
1. `SHADER_FEATURE_INFO_FLAG` - also defines corresponding `DXIL_SHADER_COMPUTED_FLAG` bit for ones with 1:1 correspondence (not requiring that separate declaration, like it is now), otherwise -1.  If -1, we need another way to compute this flag from the DXIL Shader Flags.  If we allowed an expression to be supplied (perhaps with a different variant of the macro), it could be computed using that expression.  This could solve the min-precision and low-precision feature flag cases.
2. `DXIL_SHADER_OPTION_FLAG` - option flag, not computed from usage.  Copied from module when initializing flags for computation per-function, compared when checking linking compatibility, etc...  Flags could be module scope or per-function or entry point scope if needed.
3. `DXIL_SHADER_COMPUTED_FLAG` - for additional computed flags that do not have 1:1 correspondence to `SHADER_FEATURE_INFO_FLAG`.

Potential definition for the `SHADER_FEATURE_INFO_FLAG` variant computed from DXIL shader flags with expression:
```
// SHADER_FEATURE_FLAG2(bit offset for the shader info flag, expression based on DXIL shader flags, name, description)
#ifndef SHADER_FEATURE_INFO_FLAG2
#define SHADER_FEATURE_INFO_FLAG2(FeatureInfoBit, ShaderFlagExpr, FlagName, Str) \
  SHADER_FEATURE_INFO_FLAG(FeatureInfoBit, -1, FlagName, Str)
#endif
```

Maybe these should all be defined in a TableGen file instead.  What do you think about that?

Comment (llvm-beanz) [reply llvm/include/llvm/BinaryFormat/DXContainerConstants.def]: I've been arguing against the phrase "feature info" because it seems nonsensical to me. The word "info" in that conveys no meaning and just adds an extra word that seems out of place. It really feels like that name came about because we use four characters for the part names and needed to fill space rather than a concise and meaningful term.

I don't think we should use TableGen for this. Writing a TableGen backend for this adds a lot more complexity than we need for these fairly straightforward expansions.

File: llvm/lib/Target/DirectX/DXContainerGlobals.cpp [hunk 15]
```diff
@@ -53,21 +53,24 @@ class DXContainerGlobals : public llvm::ModulePass {
...
   appendToCompilerUsed(M, Globals);
   return true;
 }
 
-GlobalVariable *DXContainerGlobals::getShaderFlags(Module &M) {
-  const uint64_t Flags =
-      (uint64_t)(getAnalysis<ShaderFlagsAnalysisWrapper>().getShaderFlags());
+GlobalVariable *DXContainerGlobals::getFeatureFlags(Module &M) {
+  const uint64_t FeatureFlags =
+      (uint64_t)(getAnalysis<ShaderFlagsAnalysisWrapper>()
```
Comment (farzonl) [inline llvm/lib/Target/DirectX/DXContainerGlobals.cpp:65]: `static_cast<uint64_t>`?

File: llvm/lib/Target/DirectX/DXILShaderFlags.cpp [hunk 16]
```diff
@@ -51,7 +51,12 @@ void ComputedShaderFlags::print(raw_ostream &OS) const {
   if (FlagVal == 0)
     return;
   OS << "; Note: shader requires additional functionality:\n";
-#define SHADER_FEATURE_FLAG(bit, FlagName, Str)                                \
+#define SHADER_FEATURE_FLAG(FeatureBit, DxilModuleNum, FlagName, Str)          \
+  if (FlagName)                                                                \
+    OS << ";       " Str "\n";
```
Comment (farzonl) [inline llvm/lib/Target/DirectX/DXILShaderFlags.cpp:56]: i'm not sure what the llvm equivalent is but seems like we should be doing something like std::setw.

File: llvm/include/llvm/BinaryFormat/DXContainerConstants.def [hunk 17]
```diff
@@ -9,47 +9,64 @@ CONTAINER_PART(OSG1)
 CONTAINER_PART(PSG1)
 
 #undef CONTAINER_PART
-#endif 
+#endif
```
Comment (farzonl) [inline llvm/include/llvm/BinaryFormat/DXContainerConstants.def:12]: maybe a comment here like // CONTAINER_PART

**Review by farzonl — APPROVED** [review]

File: llvm/lib/Target/DirectX/DXILShaderFlags.cpp (hunk 16 above)
Comment (python3kgae) [reply llvm/lib/Target/DirectX/DXILShaderFlags.cpp]: Fixed with raw_ostream::indent.

File: llvm/lib/Target/DirectX/DXContainerGlobals.cpp (hunk 15 above)
Comment (python3kgae) [reply llvm/lib/Target/DirectX/DXContainerGlobals.cpp]: Updated.

Comment (python3kgae) [reply llvm/lib/Target/DirectX/DXContainerGlobals.cpp]: Updated.

File: llvm/include/llvm/BinaryFormat/DXContainerConstants.def (hunk 17 above)
Comment (python3kgae) [reply llvm/include/llvm/BinaryFormat/DXContainerConstants.def]: Added.

File: llvm/lib/Target/DirectX/DXILShaderFlags.cpp (hunk 16 above)
Comment (python3kgae) [reply llvm/lib/Target/DirectX/DXILShaderFlags.cpp]: Fixed with raw_ostream::indent.
